osbrain/tests/test_agent.py

```diff
@@ -105,6 +105,88 @@ def test_agent_shutdown(nsproxy):
     assert 'a0' not in nsproxy.list()
 
 
+def test_close_socket_agent_shutdown(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    is shut down
+    """
+    agent = run_agent('name')
+    address = agent.bind('PUSH')
+    agent.shutdown()
+    assert not address.address.exists()
+
+
+def test_close_socket_agent_kill(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    is killed
+    """
+    agent = run_agent('name')
+    address = agent.bind('PUSH')
+    agent.kill()
+    assert not address.address.exists()
+
+
+def test_close_socket_agent_blocked(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed when a blocked
+    agent is requested to shut down
+    """
+    def log(agent, msg):
+        time.sleep(3)
+        return msg
+
+    alice = run_agent('alice')
+    bob = run_agent('bob')
+
+    addr = alice.bind('REP', alias='main', handler=log)
+    bob.connect(addr, alias='main')
+
+    Timer(1, nsproxy.shutdown).start()
+    bob.send('main', 'Hello world')
+    rep = bob.recv('main')
+    print(rep)
+
+    assert not addr.address.exists()
+
+
+def test_close_socket_agent_crash(nsproxy):
+    """
+    Check that the socket is closed and the socket file removed after the agent
+    crashes
+    """
+    agent = run_agent('agent')
+    addr = agent.bind('PUSH', 'main')
+    with pytest.raises(RuntimeError):
+        agent.raise_exception()
+
+    assert not addr.address.exists()
```

Comment:
I am not sure we should be closing all sockets after an agent crashes. This prevents the user from being able to recover the agent very easily (i.e.: just re-running the main thread).

The idea I had in mind was rather to call a `nsproxy.shutdown()` after the crash (and/or an `agent.kill()`) and that should clean the sockets just fine.

But maybe we should close the sockets when a crash occurs straight away...

Reply:
Are we assuming agents crash frequently? I believe that is an error situation where the execution should stop because  the user has an error in the code somewhere. Am I wrong?  Does this happen frequently enough for us to keep the connections alive?

Reply:
They should not, for sure. The reason we leave the agent alive after a runtime exception is for debugging purposes (i.e.: the user can create a proxy and see the agent state when the crash occured). Not sure if having the sockets open could help for debugging as well... probably not.

Reply:
Now that you mention it,  it probably would help to keep them open. If you intend to revive the agent after a crash (as long as "crashing" means raising an exception and not any more serious problems), keeping the sockets alive would make things much easier. After all, as long as the agent is not unregistered from the nameserver, calling `ns.shutdown()` will close the socket anyways.

Reply:
@ocaballeror You can keep the test though. Only instead of checking the socket just after the crash, calll `ns.shutdown()` before (you'll need to create the nameserver manually instead of using `nproxy` fixture).